### PR TITLE
chore: part 1/N of refactoring vector index into separate crate

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

File: rust/lance/src/dataset.rs [hunk 1]
```diff
@@ -382,8 +382,8 @@ impl Dataset {
                 let m = d.manifest.as_ref();
                 if schema != m.schema {
                     return Err(Error::SchemaMismatch {
-                        original: m.schema.clone(),
```
Comment (eddyxu) [inline rust/lance/src/dataset.rs:385]: Gonna re-eanble these after we cleanly migrate `schema` to `lance-core`

File: rust/lance/src/io/commit/dynamodb.rs [hunk 2]
```diff
@@ -123,7 +126,11 @@ impl DynamoDBExternalManifestStore {
             .describe_table()
             .table_name(table_name)
             .send()
-            .await?;
+            .await
+            .map_err(|e| Error::IO {
+                message: format!("dynamodb error: {}", e,),
```
Comment (chebbyChefNEQ) [inline rust/lance/src/io/commit/dynamodb.rs:131]: can you keep the source error?

dynamodb formatter hides the read error if we only print `{}`

Comment (eddyxu) [reply rust/lance/src/io/commit/dynamodb.rs:131]:
```
349 |             .map_err(|e| Error::IO {
    |                       - binding `e` declared here
350 |                 message: format!("dynamodb error: {}, source: {:?}", e, e.into_source()),
    |                          -----------------------------------------------^---------------
    |                          |                                           |  |
    |                          |                                           |  move out of `e` occurs here
    |                          |                                           borrow of `e` occurs here
    |                          borrow later used here
```

did i do sth wrong here?

Comment (eddyxu) [reply rust/lance/src/io/commit/dynamodb.rs:131]: We will have the `From<SDKError> for Error` back once this code is migrated to `lance-core`

File: rust/lance-index/src/vector/ivf.rs [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+use snafu::{location, Location};
+use tracing::instrument;
+
+use super::PART_ID_COLUMN;
+use crate::vector::transform::Transformer;
+
+/// IVF - IVF file partition
+///
+#[derive(Debug, Clone)]
+pub struct Ivf {
```
Comment (westonpace) [inline rust/lance-index/src/vector/ivf.rs:44]: This is duplicating the `Ivf` struct that's in the `lance` crate right? Or was this supposed to be a move?

**Review by westonpace — APPROVED** [review]
I think these changes are good.  Will defer to @chebbyChefNEQ on the impact of the dynamo errors.  I think the arrow schema error being temporarily disabled is probably fine.

File: rust/lance-index/src/vector/ivf.rs (hunk 3 above)
Comment (eddyxu) [reply rust/lance-index/src/vector/ivf.rs:44]: I am going to move the IVF model from lance to here. Wanted to get this in before it becomes too big.

**Pull request merged**